lib.modules: Add hint when using `config` in `imports`

49016 opened the pull request:
Minor change: Add a hint for when this happens. I've been asked a couple of times why this gets an infinite recursion; why not add a helpful error message to explain it automatically?

```
nix-repl> (lib.evalModules { modules = [ (
  {config, ...}: {
    imports = if config.x then ["a"] else ["b"];
  }
) ]; }).config

error:
       … while evaluating the attribute 'config'
         at /home/lexi/git/nixpkgs/lib/modules.nix:360:9:
          359|         options = checked options;
          360|         config = checked (removeAttrs config [ "_module" ]);
             |         ^
          361|         _module = checked (config._module);

       … while calling the 'seq' builtin
         at /home/lexi/git/nixpkgs/lib/modules.nix:360:18:
          359|         options = checked options;
          360|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          361|         _module = checked (config._module);

       … while evaluating the module argument `config' in ":anon-1":

       … If you get an infinite recursion here, you probably reference `config`
         in `imports`. This is not supported; consider using mkEnableOption.

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: infinite recursion encountered
       at /home/lexi/git/nixpkgs/lib/modules.nix:257:21:
          256|                     options
          257|                     config
             |                     ^
          258|                     specialArgs

nix-repl> 
```

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

49016 committed:
lib.modules: Add hint when using `config` in `imports`

roberth commented:
This would be valuable info, but I'm concerned that it will show up in cases where it shouldn't.
I've contemplated a new primop that would make this behavior conditional on whether some argument is a `tBlackhole`, i.e. whether it's involved in the current set of value/thunk objects that is under evaluation. I believe that would make the message reliable, and we'd also have the opportunity to make it pop out to the user instead of hiding it in the trace.
- https://github.com/NixOS/nix/issues/8629

49016 commented:
I tested it with a couple of other cases, for example when recursing on `config.x = config.x` and it does not show up there as at that point `applyModuleArgs` is done running already; this only happens when the error is in evaluating the arguments themselves, so it seems to be fine here.

(Also sorry for the 2nd force push now, somehow the `~/git/nixpkgs]$ nix-shell --run treefmt` disagrees with the GH Action, I just manually applied the patch now lol)

roberth committed:
lib.modules: Test infinite recursion hint

We don't want it to occur in the trace of any unrelated errors.

roberth committed:
lib.modules: Generalize the import hint to _module.args

roberth committed:
lib.modules: Make _module.args evaluation explicit in trace

roberth committed:
lib.modules: Adjust error message

- Lower case error trace for consistency
- Be more explicit about the condition under which the hint applies,
  and the resolution.

roberth commented:
I think you're right.
My primop idea would allow the message to be a single, coherent one, but after generalizing your solution, I find that plain `addErrorContext` produces a remarkably good result:

Here's an import from `_module.args.custom`:

```

       … while evaluating the module argument `custom' in "/home/user/src/nixpkgs-master/lib/tests/modules/import-custom-arg.nix":

       … noting that argument custom is not externally provided, and querying `_module.args` instead, requiring `config`

       … if you get an infinite recursion here, you probably reference `config` in `imports`. This is not possible. If you are trying to achieve a conditional behavior dependent on `config`, consider importing unconditionally, and using `mkEnableOption` and `mkIf` to control its effect.

       error: infinite recursion encountered
```

The test helps to ensure that we don't provide this context where it's not relevant or accurate.
I've also made a few attempts to get it to trigger incorrectly but failed.
I'd gladly release this change and correct it if needed.

I believe we could further generalize this to `options`, but that could be a follow-up, and it doesn't have nearly the same impact.

Do my changes look ok to you?

roberth commented:
> ofborg-eval — Failed to merge

No further details, yet I can merge locally. Let's give it a friendly kick

@ofborg eval

hsjobeki commented on lib/modules.nix:
```diff
@@ -254,11 +254,11 @@ let
                     ;
                   _class = class;
                   _prefix = prefix;
+                  config = addErrorContext "if you get an infinite recursion here, you probably reference `config` in `imports`. This is not possible. If you are trying to achieve a conditional behavior dependent on `config`, consider importing unconditionally, and using `mkEnableOption` and `mkIf` to control its effect." config;
```
```suggestion
                  config = addErrorContext "if you get an infinite recursion here, you probably reference `config` in `imports`. If you are trying to achieve a conditional import behavior dependent on `config`, consider importing unconditionally, and using `mkEnableOption` and `mkIf` to control its effect." config;
```

Using config in imports is possible in general. But its not possible to do conditional imports where the condition depends on config. Thats two different statements

hsjobeki reviewed (commented):
Overall this seems lgtm to me, error improvement is always nice. :+1:

roberth committed:
lib.modules: Improve infinite recursion hint

hsjobeki: Using config in imports is possible in general. But its not possible to do conditional imports where the condition depends on config. Thats two different statements.

Co-authored-by: Johannes Kirschbauer <hsjobeki+github@gmail.com>
